### PR TITLE
annotations: Expands list of health-probe related ingress annotations 

### PR DESCRIPTION
**Pull request by polasekr**
## Checklist
- [X] The title of the PR is clear and informative
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description
It expands list of recognized ingress annotations with following additions:
- appgw.ingress.kubernetes.io/health-probe-hostname
- appgw.ingress.kubernetes.io/health-probe-port
- appgw.ingress.kubernetes.io/health-probe-path
Those annotations allow user override health probe settings with values from annotations.

## Fixes
https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1048

### CODE REVIEW

**Comment by akshaysngupta** [comment]
@polasekr can you also add an annotation for supporting allowed status code ?

**Comment by polasekr** [comment]
As suggested I added annotation `health-probe-status-codes` plus 3 others
- health-probe-interval 
- health-probe-timeout 
- health-probe-unhealthy-treshold

**Review by akshaysngupta — CHANGES_REQUESTED** [review]

File: docs/annotations.md [hunk 1]
```diff
@@ -478,4 +485,228 @@ spec:
...
+```
+
+## Health Probe Interval
+
+This annotation sets AGW health probe interval. By default, if backend container running service with liveliness probe of type `HTTP GET` defined, interval in liveliness probe definition is also used as a interval for health probe. However annotation `appgw.ingress.kubernetes.io/health-probe-interval` overrides it with its value.
+
+### Usage
+
+```yaml
+appgw.ingress.kubernetes.io/health-probe-path: <URI path>
```
[the remark this answers is not in the document]
Comment (polasekr) [reply docs/annotations.md:628]: corrected

File: docs/annotations.md [hunk 2]
```diff
@@ -478,4 +485,228 @@ spec:
...
+```
+
+## Health Probe Timeout
+
+This annotation allows specifically define timeout for AGW health probe. By default, if backend container running service with liveliness probe of type `HTTP GET` defined, timeout defined in liveliness probe definition is also used for health probe. However annotation `appgw.ingress.kubernetes.io/health-probe-timeout` overrides it with its value.
+
+### Usage
+
+```yaml
+appgw.ingress.kubernetes.io/health-probe-path: <URI path>
```
[the remark this answers is not in the document]
Comment (polasekr) [reply docs/annotations.md:659]: corrected

File: docs/annotations.md [hunk 3]
```diff
@@ -478,4 +485,228 @@ spec:
...
+```
+
+## Health Probe Unhealthy Treshold
+
+This annotation allows specifically define target unhealthy thresold for AGW health probe. By default, if backend container running service with liveliness probe of type `HTTP GET` defined , threshold defined in liveliness probe definition is also used for health probe. However annotation `appgw.ingress.kubernetes.io/health-probe-unhealthy-treshold` overrides it with its value.
+
+### Usage
+
+```yaml
+appgw.ingress.kubernetes.io/health-probe-path: <URI path>
```
[the remark this answers is not in the document]
Comment (polasekr) [reply docs/annotations.md:690]: corrected

File: docs/annotations.md [hunk 4]
```diff
@@ -478,4 +485,228 @@ spec:
...
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: go-server-ingress-bkprefix
+  namespace: test-ag
+  annotations:
+    kubernetes.io/ingress.class: azure/application-gateway
+    appgw.ingress.kubernetes.io/health-probe-unhealthy-treshold: "5"
```
[the remark this answers is not in the document]
Comment (polasekr) [reply docs/annotations.md:703]: corrected

File: docs/annotations.md [hunk 5]
```diff
@@ -478,4 +485,228 @@ spec:
...
+```
+
+## Health Probe Status Codes
+
+This annotation defines healthy status codes returned by the health probe. The values are comma seperated list of individual status codes or ranges defined as `<start of the range>-<end of the range>`.
+
+### Usage
+
+```yaml
+appgw.ingress.kubernetes.io/health-probe-path: <URI path>
```
[the remark this answers is not in the document]
Comment (polasekr) [reply docs/annotations.md:597]: corrected

File: pkg/annotations/ingress_annotations_test.go [hunk 6]
```diff
@@ -105,6 +112,105 @@ var _ = Describe("Test ingress annotation functions", func() {
...
+			Expect(actual).To(Equal(int32(0)))
+		})
+		It("returns hostname", func() {
+			actual, err := HealthProbeTimeout(ing)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(actual).To(Equal(int32(10)))
+		})
+	})
+
+	Context("test health-probe-unhealthy-treshold", func() {
```
[the remark this answers is not in the document]
Comment (polasekr) [reply pkg/annotations/ingress_annotations_test.go:200]: corrected

File: pkg/annotations/ingress_annotations_test.go [hunk 7]
```diff
@@ -105,6 +112,105 @@ var _ = Describe("Test ingress annotation functions", func() {
...
+	})
+
+	Context("test health-probe-status-codes", func() {
+		It("returns error when ingress has no annotations", func() {
+			ing := &v1beta1.Ingress{}
+			actual, err := HealthProbeStatusCodes(ing)
+			Expect(err).To(HaveOccurred())
+			Expect(actual).Should(BeEmpty())
+		})
+		It("returns hostname", func() {
```
[the remark this answers is not in the document]
Comment (polasekr) [reply pkg/annotations/ingress_annotations_test.go:164]: changed to "returns expected status codes"

File: pkg/appgw/health_probes.go [hunk 8]
```diff
@@ -158,10 +158,58 @@ func (c *appGwConfigBuilder) generateHealthProbe(backendID backendIdentifier) *n
 		}
 	}
 
+	// backend protocol must match http settings protocol
+	backendProtocol, err = annotations.BackendProtocol(backendID.Ingress)
```
[the remark this answers is not in the document]
Comment (polasekr) [reply pkg/appgw/health_probes.go:162]: Whole section of code starting line 112 and ending at 120 has been removed

File: pkg/appgw/health_probes_test.go [hunk 9]
```diff
@@ -183,4 +185,73 @@ var _ = Describe("configure App Gateway health probes", func() {
...
+	Context("ensure that annotation overrides defaults for health probe", func() {
+
+		annotationHpHostname := "myhost.mydomain.com"
+		annotationHpPort := int32(8080)
+		annotationHpPath := "/healthz"
+		annotationHpCodes := "200-399,401"
+		annotationHpInterval := int32(15)
+		annotationHpTimeout := int32(10)
+		annotationHpTreshold := int32(3)
+		StatusCodes := strings.Split(annotationHpCodes, ",")
```
[the remark this answers is not in the document]
Comment (polasekr) [reply pkg/appgw/health_probes_test.go:197]: changed

File: docs/annotations.md [hunk 10]
```diff
@@ -22,42 +22,14 @@ For an Ingress resource to be observed by AGIC it **must be annotated** with `ku
...
-| [appgw.ingress.kubernetes.io/override-frontend-port](#override-frontend-port) | `string` |   |   |
+| [appgw.ingress.kubernetes.io/health-probe-hostname](#health-probe-hostname) | `string` |  `nil` |   |
+| [appgw.ingress.kubernetes.io/health-probe-port](#health-probe-port) | `int32` | `nil`  |   |
+| [appgw.ingress.kubernetes.io/health-probe-path](#health-probe-path) | `string` | `nil`  |   |
+| [appgw.ingress.kubernetes.io/health-probe-status-codes](#health-probe-status-codes) | `[]string` | `nil`  |   |
+| [appgw.ingress.kubernetes.io/health-probe-interval](#health-probe-interval) | `int32` | `nil`  |   |
+| [appgw.ingress.kubernetes.io/health-probe-timeout](#health-probe-timeout) | `int32` | `nil`  |   |
+| [appgw.ingress.kubernetes.io/health-probe-unhealthy-threshold](#health-probe-unhealthy-threshold) | `int32` | `nil`  |   |
 
-## Override Frontend Port
```
Comment (akshaysngupta) [inline docs/annotations.md:27]: can undo removing "Override Frontend Port" doc

**Comment by akshaysngupta** [comment]
@polasekr PR looks good.
Added one comment to correct the doc.

File: docs/annotations.md (hunk 10 above)
Comment (polasekr) [reply docs/annotations.md:27]: The change in the documentation related to "Override frontend port" has been reverted

**Review by akshaysngupta — APPROVED** [review]
LGTM
Thanks for contributing. You can try your changes.

**Comment by akshaysngupta** [comment]
@polasekr You can try your changes tomorrow using [nightly build](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/developers/nightly.md).

**Comment by oofnikj** [comment]
Hi,

When can we expect a new version to be released incorporating this PR?

Thank you

**Comment by andreaslarfors** [comment]
Also waiting for these features to be available. Cannot utilize App Gateway until then.

**Comment by oofnikj** [comment]
Hi @andreaslarfors, we are using the nightly build until the next release is cut.
@akshaysngupta this PR was merged over two months ago. A new versioned release would be most helpful.

**Comment by picardsrcd** [comment]
Has this been put in a release?

**Pull request merged**